server/src/main/java/org/elasticsearch/index/fieldvisitor/IgnoredSourceFieldLoader.java

```diff
@@ -0,0 +1,142 @@
+
+    IgnoredSourceFieldLoader(StoredFieldsSpec spec) {
+        Set<String> potentialFieldsInIgnoreSource = new HashSet<>();
+        for (String requiredStoredField : spec.requiredStoredFields()) {
+            if (requiredStoredField.startsWith(IgnoredSourceFieldMapper.NAME)) {
+                String fieldName = requiredStoredField.substring(IgnoredSourceFieldMapper.NAME.length());
+                potentialFieldsInIgnoreSource.addAll(FallbackSyntheticSourceBlockLoader.splitIntoFieldPaths(fieldName));
+            }
+        }
+        this.potentialFieldsInIgnoreSource = potentialFieldsInIgnoreSource;
+    }
+
+    @Override
+    public LeafStoredFieldLoader getLoader(LeafReaderContext ctx, int[] docs) throws IOException {
+        var reader = sequentialReader(ctx);
+        var visitor = new SFV(potentialFieldsInIgnoreSource);
+        return new LeafStoredFieldLoader() {
+
+            private int doc = -1;
+
+            @Override
+            public void advanceTo(int doc) throws IOException {
+                if (doc != this.doc) {
+                    visitor.reset();
+                    reader.accept(doc, visitor);
+                    this.doc = doc;
+                }
+            }
+
+            @Override
+            public BytesReference source() {
+                return null;
+            }
+
+            @Override
+            public String id() {
+                return null;
+            }
+
+            @Override
+            public String routing() {
+                return null;
+            }
+
+            @Override
+            public Map<String, List<Object>> storedFields() {
+                return Map.of(IgnoredSourceFieldMapper.NAME, visitor.values);
+            }
+        };
+    }
+
+    @Override
+    public List<String> fieldsToLoad() {
+        return List.of(IgnoredSourceFieldMapper.NAME);
+    }
+
+    static class SFV extends StoredFieldVisitor {
+
+        boolean done;
+        final List<Object> values = new ArrayList<>();
+        final Set<String> potentialFieldsInIgnoreSource;
+
+        SFV(Set<String> potentialFieldsInIgnoreSource) {
+            this.potentialFieldsInIgnoreSource = potentialFieldsInIgnoreSource;
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            if (done) {
+                return Status.STOP;
+            } else if (IgnoredSourceFieldMapper.NAME.equals(fieldInfo.name)) {
+                return Status.YES;
+            } else {
+                return Status.NO;
+            }
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            var result = IgnoredSourceFieldMapper.decodeIfMatch(value, potentialFieldsInIgnoreSource);
+            if (result != null) {
+                // TODO: can't do this in case multiple entries for the same field name. (objects, arrays etc.)
+                // done = true;
```

Comment:
Note this doesn't work in the case that multiple ignored source entries exist for the same field. This is possible with nested objects and arrays that share the same name. 

Looks like the best way to achieve is for each ignored source entry to use a unique stored field name and not reuse `_ignored_source` as stored field name. This shouldn't be too difficult, given that each ignored source entry is already stored separately, but just with the same name.